File: python/040. Combination Sum II.py
class Solution(object):
    def dfs(self, candidates, target, idx, cur, res):
        if target == 0: res.append(copy.copy(cur))
        if idx >= len(candidates): return
        while idx < len(candidates):
            val = candidates[idx]
            if val > target:
                break
            cur.append(val)
            self.dfs(candidates, target - val, idx + 1, cur, res)
            cur.pop()
            while idx < len(candidates) and candidates[idx] == val:
                idx += 1
        return
    
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        n = len(candidates)
        cur = []
        res = []
        candidates.sort()
        self.dfs(candidates, target, 0, cur, res)
        return res

File: C++/040. Combination Sum II.cpp
class Solution {
private: 
    void SearchSum(vector<vector<int>> &res, vector<int>& candidates, vector<int>& sol, int target, int index)
    {
        for(int j=index;candidates[j]<=target && j < candidates.size();j++)
        {
            if (candidates[j]<=target)
            {
                sol.push_back(candidates[j]);
                if(candidates[j]==target)
                {
                    res.push_back(sol);
                    sol.pop_back();
                    return;
                }
                else
                {
                    SearchSum(res, candidates, sol, target-candidates[j], j+1);
                    sol.pop_back();
                }
            }
            else{break;}
            while(j+1<candidates.size()&&candidates[j]==candidates[j+1]){j++;}
        }
        return;
    }

public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> sol;
        vector<vector<int>> res;
        int size = candidates.size();
        if (size<=0){return res;}
        sort(candidates.begin(),candidates.end());
        SearchSum(res, candidates, sol, target, 0);
        return res;
    }
};
